### ISSUE TITLE
Option to generate URL previews at the receiving client, not the server

### ISSUE THREAD

**Issue by carlos22** [vector-im/element-meta#1138]

**Comment by ara4n** [comment]
Alternatively, we could make it the sending client's problem (as per vector-im/element-meta#1139).  It's an interesting question as to whether the preview should be telling the user what they will see if they click on the page (whilst leaking their IP, and hammering the server)... or telling the user what the sender intended the preview to be (as per vector-im/element-meta#1139).

Unsure how to proceed here, really.

**Comment by teohhanhui** [comment]
Generating on the sender side is the right thing to do.

Keybase does the same: https://book.keybase.io/docs/chat/link-previews

**Comment by kevincox** [comment]
That link seems to show that keybase does it client side.

**Comment by teohhanhui** [comment]
> That link seems to show that keybase does it client side.

Yes, server-side is simply not an option when it's end-to-end encrypted.

I'm actually backing up what you've said above:

> I think the best solution is to make the sender generate a preview.

**Comment by kevincox** [comment]
Oh sorry, I misread "sender-side" as "server-side".

**Comment by HarHarLinks** [comment]
A common talking point and strong counter-argument to client-side previews appears to be
> People don't like leaking their IP address

Personally, I think this use case described by kevincox applies more often:
> For example I am often sharing "secret links" in E2EE chats. I am not too worried about leaking my IP (at least in some rooms) but I would rather not share that URL with the server.

This applies only limitedly (?? can you say this in english?) when you both are hosted on a trustworthy homeserver, e.g. all conversation parties are self-hosting. For the time being this is the rarer scenario, however you could argue that privacy concerned people would tend to do this anyway if they are able to.

In my opinion the strongest argument is that generating server-side takes any control out of your hands. Even if it is your own server, it will get associated with opening the link. When you are concerned about your IP leaking, it is reasonable to assume that you have taken steps for your whole PC, as you likely use more services than just element. For example you are likely using a VPN you trust. It seems reasonable that you are going to click a link you received eventually. In this case it would be better and safer to generate the preview from your own client.

I think https://github.com/vector-im/element-android/issues/3038 makes an interesting point: "[room] Administrators get to decide if the users of a room can individually enable/disable URL-Previews for their client. (defaulting to: disabled each)". In the context of this issue this could mean that room admins may disable server-side preview generation for all room participants.

**Comment by t3chguy** [comment]
Client side generation in the web can be problematic due to CORS preventing access to other domains unless they have `ACAO: *`

**Comment by Doomsdayrs** [comment]
Signal generates previews client side, and there seems to be no issues with that

**Comment by t3chguy** [comment]
@Doomsdayrs Signal doesn't look to be a webapp?

**Comment by Doomsdayrs** [comment]
> @Doomsdayrs Signal doesn't look to be a webapp?

It is not, but the android app shows that it is possible to get such a system done. Load the embed on user side and the user sends it out

**Comment by t3chguy** [comment]
Only Web apps are bound by [ACAO headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin), they're enforced by web browsers, not by native applications.

**Comment by K4LCIFER** [comment]
I would also much prefer that URL thumbnails be generated locally instead of on the server. To me, it makes little practical sense to have it generated on the server, not to mention the security implications of it being generated on the server. I presume that if it is being generated on the server, then the server knows that that specific user is requesting a thumbnail for that specific link contained within a specific room; however, if the URLs are generated client side, its just a normal websearch from the user that the server knows nothing about. I don't think that there should be any option to toggle it server-side. I think that all URL thumbnails should either be generated locally, or not at all (by default, I think URL thumbnails should be off).

**Comment by maltfield** [comment]
+1 for client-side (either sender- or receiver-generated)

I'll just add that many of the more privacy-concerned users are likely already forcing all of their traffic through a shared proxy server (eg VPN or Tor) anyway. So "leaking" their IP isn't really a concern for many users.

I do think there should be some client-side cache, though. Once the URL is fetched once by the user, they shouldn't be sending the query off to regenerate the preview the next time they reboot their device and launch element again (I guess this concern would be elegantly solved by generating the preview sender-side).

**Comment by HarHarLinks** [comment]
~~client side generation would be a way to generate previews for URLs that can only really be accessed after having logged in (by reusing the browser session, if it is a web client). a use case would be github issues on secret repos, though today we have bots to work around that.~~

**Comment by kevincox** [comment]
> by reusing the browser session, if it is a web client

This won't work for the general case because if a site can generate a preview for another site's private data that is a cross-site information leak. Browsers also are starting to block cross-site cookies by default, so even sites with this vulnerability will not work in the near future.

If you did want previews for non-public information the options I can see are:

1. Use a privileged client to pull your authentication info out of your browser (or have the user configure cookies/auth manually)
3. Have a bot with access attach previews to messages.
4. Teach your client to generate this preview with a "special" method such as using a configured GitHub token to fetch preview data from the API.
5. Embed some sort of token in the URL that allows previews.

None of these are general solutions. 1 requires a privileged client or manual configuration, 2 requires a bot with manual configuration, 3 requires manual auth and per-site code and 4 requires website support and user-action to get this custom link.

For private information I think that 2 (bot) will be the easiest option forward. This is basically how it works in Slack. You sign into the "Slack App" and then it can generate and attach previews to messages using your authentication information. However this does mean that we will want the option for other users to attach previews to other users' messages.

Having different users attach previews to your messages is definitely flexible but also has lots of abuse potential. So likely there would need to be permission controls and probably UX.

1. Permission level to control who can attach previews to messages. (Probably with a special case for attaching previews to their own messages).
2. Some sort of client UX to indicate who generated which preview.

I suspect that we want 2 no matter what. But we will also need 1 if people can attach previews to other user's messages. IDK if this should be room permissions, or configured by the client. I can see use cases for both. I would lean towards client-side because then each user is control of which previews they see. However for large rooms it may be best to add at least some default rules to avoid confusing preview spam.

**Comment by Mikaela** [comment]
Generating previews on client would be scary considering how much Element's user-agent tells the Matrix server already https://github.com/element-hq/element-meta/discussions/2309.

> I'll just add that many of the more privacy-concerned users are likely already forcing all of their traffic through a shared proxy server (eg VPN or Tor) anyway.

Tor is seldom an option judging by #200 and [Matrix isn't that great for privacy](https://aminda.eu/blog/english/2021/08/03/matrix-perfect-privacy-not.html#you-mentioned-privacy).

**Comment by drzewiec** [comment]
Just to throw my $.02 in... generating previews on the client side may be necessary to get them to work at all in some cases. I've been playing around with running a homeserver in Digital Ocean (my usual provider of choice), and half the URLs I try to test preview with return nothing at all, because the host is blacklisting DO's IP.

I get why they do it, of course. But I also don't really know what recourse I have - from experience with AWS their IPs get blocked too, and I would expect most any hosting provider is going to be the same. So without the ability to generate URL previews from the client end, I'm kind of stuck in terms of being able to take advantage of this feature.

**Comment by maltfield** [comment]
@drzewiec interesting point, but the opposite could be true too: where users that must use commercial VPNs (because of the risks associated with the country that they live-in) and their shared VPN IP address might be blocked (or proxied by some front like CloudFlare or Akamai that present a CAPTCHA challenge that would break previews).

But this is a good argument to do it client-side twice: [1] first try it at the sender's side and [2] (if it arrives without a preview) later try at the receiver's side.
